libyul/optimiser/FullInliner.cpp

```diff
@@ -190,7 +188,7 @@ vector<Statement> InlineModifier::performInline(Statement& _statement, FunctionC
 		if (_value)
 			varDecl.value = make_unique<Expression>(std::move(*_value));
 		else
-			varDecl.value = make_unique<Expression>(zero);
+			varDecl.value = make_unique<Expression>(Literal{{}, LiteralKind::Number, YulString{"0"}, {}});
```

Comment:
Doesn't this create a dangling reference inside yulstringrepo due to no reset callback being registered?

Reply:
No, it is stored inside the AST which is assumed to not live beyond a compiler run.

Reply:
Okay, I must admit I don't understand YulString storage very well. Let me summarize what I think this code does (please correct me if I am wrong)

- Yul strings are constructed here

https://github.com/ethereum/solidity/blob/6e9bb2c07304deba09a36b31cde49fa5895d66c2/libyul/YulString.h#L96

- At the time of construction, a handle is created for the said string via the `YulStringRepository` singleton object

https://github.com/ethereum/solidity/blob/6e9bb2c07304deba09a36b31cde49fa5895d66c2/libyul/YulString.h#L53-L66

- During the string to handle conversion, additional memory may be consumed by these two emplaces

https://github.com/ethereum/solidity/blob/7d9007dc04554710cc8b3ff7e6370503728ad332/libyul/YulString.h#L62

https://github.com/ethereum/solidity/blob/7d9007dc04554710cc8b3ff7e6370503728ad332/libyul/YulString.h#L64

- Note that these two emplaces increase the memory of the `YulStringRepository` object

- Finally, when reset on the YulStringRepo object is called, the singleton object (YulStringRepository) is reassigned a fresh copy of the object.
  - What happens to the memory consumed by the emplaces above? Are they free by `YulStringRepository`'s destructor?
  - What happens if `m_strings` (being a shared ptr) is still referenced elsewhere? Won't its memory footprint still remain?

https://github.com/ethereum/solidity/blob/7d9007dc04554710cc8b3ff7e6370503728ad332/libyul/YulString.h#L106

Reply:
Each YulStringRepository contains its own `m_strings` member. Usually, there is exactly one such repository, because we mostly only use the singleton. The `m_strings` member (and also the memory for the strings) is deallocated in its destructor. This destructor is invoked when a YulStringRepository object is destroyed or when the `m_strings` member is assigned from somewhere else. The first case happens when the program terminates (although it might be skipped) and the second case happens in the reset function.